lib/vrl/DESIGN.md

```diff
@@ -0,0 +1,269 @@
+limited in their capability to solve the required need.
+
+### Functions
+
+#### Composition
+
+- Favor function composition over single-purpose functions.
+
+- If a problem needs to be solved in multiple steps, consider adding
+  single-purpose functions for each individual step.
+
+- If useability or readability is hurt by composition, favor single-purpose
+  functions.
+
+#### Naming
+
+- Function names are lower-cased (e.g. `round`).
+
+- Multi-name functions use underscore (`_`) as a separator (e.g.
+  `encode_key_value`).
+
+- Favor explicit verbose names over terse ones (e.g.
+  `parse_aws_cloudwatch_log_subscription_message` over `parse_aws_cwl_msg`).
+
+- Functions should be preceded with their function category for organization and
+  discovery.
+
+  - Use `parse_*` for string to type decoding functions (e.g. `parse_json` and
+    `parse_grok`).
+  
+  - Use `decode_*` for string to string decoding functions (e.g.
+    `decode_base64`).
+  
+  - Use `encode_*` for string encoding functions (e.g. `encode_base64`).
+  
+  - Use `to_*` to convert from one type to another (e.g. `to_bool`).
+
+  - Use `is_*` to determine if the provided value is of a given type (e.g.
+    `is_string` or `is_json`).
+  
+  - Use `format_*` for string formatting functions (e.g. `format_timestamp` and
+    `format_number`).
+
+#### Return Types
+
+- Return boolean from `is_*` functions (e.g. `is_string`).
+
+- Return a string from `encode_*` functions (e.g. `encode_base64`).
+
+- Return an error whenever the function can fail at runtime.
+
+#### Mutability
+
+As a general rule, functions never mutate values.
+
+Favor this:
+
+```coffee
+# explicitly assign the parsed JSON to `.message`
+.message = parse_json(.message)
+```
+
+Over this:
+
+```coffee
+# mutate `.message` in place
+parse_json(.message)
+```
+
+There are exceptions to this rule (such as the `del` function), but they are
+limited, and additional exceptions should be well reasoned.
+
+#### Fallibility
+
+- A function must be marked as fallible if it can fail in any way.
+
+- Design a function with the goal of making it infallible.
+
+- But don't hide fallibility for the sake of persuing the previous rule.
+
+- A function implementation can assume it receives the argument type it has
+  defined.
+
+- `parse_*` functions should almost always error when used incorrectly.
+
+- `to_*` functions must never fail.
+
+#### Signatures
+
+- Functions can have zero or more parameters (e.g. `now()` and `assert(true)`).
+
+- Argument naming follows the same conventions as function naming.
+
+- For one or more parameters, the first parameter must be the "value" the
+  function is acting on (e.g. `parse_regex(value: <string>, pattern: <regex>)`).
+
+- The first parameter must therefore almost always be named `value`.
+
+- The exception to this is when you're dealing with an actual VRL path (e.g.
+  `del(path)`) or in special cases such as `assert`.
```

Comment:
This strikes me as it seems generally usage is `parse_regex(.message, ...)` where the `value` is extracted from the `path`

Reply:
Whether a function extracts the value from a path depends on the function implementation.

Almost always, yes, a function will resolve the path to a value, but in a few cases where we want mutability (such as `del`), the function doesn't actually resolve the path to a value, but instead asks the VRL runtime to _delete the value at the given path_.

Reply:
👍 just something that came to mind as I was pondering the distinction between our "first parameters"

Reply:
Yep, I agree it can be confusing at first, especially as a developer. For users, it should (hopefully) result in a "it just works" situation.